mplc/scenario.py

```diff
@@ -724,66 +734,16 @@ def compute_batch_sizes(self):
-                logger.debug(
-                    f"  ... Corrupting (by randomizing the labels ) the whole dataset of partner #{partner.id}."
-                    f"Dirichlet distribution matrix available"
-                )
-                partner.random_labels(1.0)
-            elif self.corrupted_datasets[partner_index][0] == "corrupted":
-                logger.debug(
-                    f"   ... Corrupting (by offsetting labels) {self.corrupted_datasets[partner_index][1] * 100} \
-                            percent of the data of partner #{partner.id}"
-                )
-                partner.corrupt_labels(self.corrupted_datasets[partner_index][1])
-            elif self.corrupted_datasets[partner_index][0] == "shuffled":
-                logger.debug(
-                    f"   ... Corrupting (by shuffling labels) {self.corrupted_datasets[partner_index][1] * 100} \
-                            percent of the data of partner #{partner.id}"
-                )
-                partner.shuffle_labels(self.corrupted_datasets[partner_index][1])
-            elif self.corrupted_datasets[partner_index][0] == 'permuted':
-                logger.debug(
-                    f"  ... Corrupting (by permuting the labels ) {self.corrupted_datasets[partner_index][1] * 100} "
-                    f"percent of the data of partner #{partner.id}."
-                    f"Permutation matrix available"
-                )
-                partner.permute_labels(self.corrupted_datasets[partner_index][1])
-            elif self.corrupted_datasets[partner_index][0] == 'random':
-                logger.debug(
-                    f"  ... Corrupting (by randomizing the labels ) {self.corrupted_datasets[partner_index][1] * 100} "
-                    f"percent of the data of partner #{partner.id}."
-                    f"Dirichlet distribution matrix available"
-                )
-                partner.random_labels(self.corrupted_datasets[partner_index][1])
-            elif self.corrupted_datasets[partner_index] == "not_corrupted":
-                pass
-            else:
-                logger.debug("Unexpected label of corruption, no corruption performed!")
-
-            logger.debug(f"   Partner #{partner.id}: done.")
+        """perform corruption on partner if needed"""
+        for partner in self.partners_list:
+            if partner.corruption_method == "duplication":
+                if not partner.duplicated_partner_id:
+                    data_volume = np.array([p.data_volume for p in self.partners_list if p.id != partner.id])
+                    ids = np.array([p.id for p in self.partners_list if p.id != partner.id])
+                    candidates = ids[data_volume >= partner.data_volume * partner.proportion_corrupted]
+                    partner.duplicated_partner_id = np.choice(candidates)
+                partner.duplicated_partner = self.partners_list[partner.duplicated_partner_id]
```

Comment:
Why can't we move this to the partner class ?

Reply:
Because the partner doesn't have access to other partner, so we need to directly pass the partner to use for duplication